SimG4Core/CustomPhysics/src/RHadronPythiaDecayer.cc

```diff
@@ -0,0 +1,351 @@
+  fillParticle(aTrack, event);
+  RHadronToConstituents(event);
+  pythia_->next();
+
+  // Add the particles from the Pythia event into the Geant particle vector
+  G4ParticleTable* particleTable = G4ParticleTable::GetParticleTable();
+  for (int i = 0; i < pythia_->event.size(); i++) {
+    // If the particle status is negative and it decays outside of the vertex, change the status to positive and do not add its decay products. If its status is negative but it decays inside the detector, skip it.
+    G4ThreeVector vProd =
+        RHadronDecayLocation +
+        G4ThreeVector(pythia_->event[i].xProd(), pythia_->event[i].yProd(), pythia_->event[i].zProd());
+    G4ThreeVector vDec = RHadronDecayLocation +
+                         G4ThreeVector(pythia_->event[i].xDec(), pythia_->event[i].yDec(), pythia_->event[i].zDec());
+    if (pythia_->event[i].status() < 0 && (worldSolid->Inside(vDec) == kOutside))
+      pythia_->event[i].statusPos();
+    else if (pythia_->event[i].status() < 0)
+      continue;
+    if (worldSolid->Inside(vProd) == kOutside)
+      continue;
+
+    G4ThreeVector displacement(pythia_->event[i].xProd(), pythia_->event[i].yProd(), pythia_->event[i].zProd());
+    G4LorentzVector p4(pythia_->event[i].px(), pythia_->event[i].py(), pythia_->event[i].pz(), pythia_->event[i].e());
+    p4 *= 1000.0;  // Convert GeV to MeV
+
+    const G4ParticleDefinition* particleDefinition =
+        particleTable->FindParticle(pythia_->event[i].id());  // Get the particle definition from the Pythia event
+    if (!particleDefinition) {
+      edm::LogWarning("SimG4CoreCustomPhysics") << "RHadronPythiaDecayer: I don't know a definition for pdgid "
+                                                << pythia_->event[i].id() << "! Skipping it...";
+      continue;
+    }
+
+    G4DynamicParticle* dynamicParticle =
+        new G4DynamicParticle(particleDefinition, p4);  // Create the dynamic particle and add it to Geant
+    particles.push_back(dynamicParticle);
+    secondaryDisplacements_.push_back(
+        displacement);  // Store the position of the secondary particle to update in RHadronPythiaDecayer::DecayIt
+  }
+}
+
+void RHadronPythiaDecayer::fillParticle(const G4Track& aTrack, Pythia8::Event& event) const {
+  // Reset event record to allow for new event.
+  event.reset();
+
+  // Get particle mass and 4-momentum.
+  double mass = aTrack.GetDynamicParticle()->GetMass() / CLHEP::GeV;
+  const G4LorentzVector g4p4 = aTrack.GetDynamicParticle()->Get4Momentum() / CLHEP::GeV;
+  Pythia8::Vec4 p4(g4p4.px(), g4p4.py(), g4p4.pz(), g4p4.e());
+
+  // Store the particle in the event record.
+  event.append(aTrack.GetDefinition()->GetPDGEncoding(), 1, 0, 0, p4, mass);
+}
+
+void RHadronPythiaDecayer::RHadronToConstituents(Pythia8::Event& event) {
+  // This code is very similar to Pythia8::RHadrons::decay(). Unfortunately, it is not possible in this scenario to use Pythia8::RHadrons::decay().
```

Comment:
Yes, unfortunately `nRHad` is a private member in `Pythia8::RHadrons`. If it were protected, one could make an inherited class and modify its value manually. We could propose to change this in the future.